Inline some functions used in copyCompactAndLinkCode

justinmichaud opened the pull request:
#### e3f2a6f91ba48a9da351d92f82cd28f37a087f07
<pre>
Inline some functions used in copyCompactAndLinkCode
<a href="https://bugs.webkit.org/show_bug.cgi?id=266362">https://bugs.webkit.org/show_bug.cgi?id=266362</a>
<a href="https://rdar.apple.com/119627634">rdar://119627634</a>

Reviewed by Mark Lam.

Inline some functions used in copyCompactAndLinkCode, since we want
to minimize register spilling there.

* Source/JavaScriptCore/assembler/ARM64Assembler.h:
(JSC::ARM64Assembler::buffer):
(JSC::ARM64Assembler::LinkRecord::from const):
(JSC::ARM64Assembler::LinkRecord::setFrom):
(JSC::ARM64Assembler::LinkRecord::to const):
(JSC::ARM64Assembler::LinkRecord::type const):
(JSC::ARM64Assembler::LinkRecord::linkType const):
(JSC::ARM64Assembler::LinkRecord::branchType const):
(JSC::ARM64Assembler::LinkRecord::setLinkType):
(JSC::ARM64Assembler::LinkRecord::condition const):
(JSC::ARM64Assembler::LinkRecord::is64Bit const):
(JSC::ARM64Assembler::LinkRecord::isThunk const):
(JSC::ARM64Assembler::LinkRecord::bitNumber const):
(JSC::ARM64Assembler::LinkRecord::compareRegister const):
* Source/JavaScriptCore/assembler/AssemblerBuffer.h:
(JSC::AssemblerBuffer::putInt):
(JSC::AssemblerBuffer::releaseAssemblerData):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::MacroAssemblerARM64):
(JSC::MacroAssemblerARM64::jumpsToLink):
(JSC::MacroAssemblerARM64::canCompact):
(JSC::MacroAssemblerARM64::computeJumpType):
(JSC::MacroAssemblerARM64::jumpSizeDelta):
(JSC::MacroAssemblerARM64::isCompactPtrAlignedAddressOffset):

Canonical link: <a href="https://commits.webkit.org/272034@main">https://commits.webkit.org/272034@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/259f2a1761dca08c17931eb72795ee71f32577cc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32069 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27547 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11358 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6363 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30759 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/7669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/27300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/6571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/27143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34299 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/26150 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/27738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/27640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/30557 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6705 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/8452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36995 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/7442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7959 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3933 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7249 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddegazio commented on Source/JavaScriptCore/assembler/AssemblerBuffer.h:
```diff
@@ -344,7 +344,7 @@ namespace JSC {
 #endif
         void putIntUnchecked(int32_t value) { putIntegralUnchecked(value); }
-        void putInt(int32_t value) { putIntegral(value); }
+        ALWAYS_INLINE void putInt(int32_t value) { putIntegral(value); }
```
Shouldn't we inline `putIntUnchecked`, `putIntegral`, `putIntegralUnchecked`, and their variants while we're at it?

ddegazio commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
```diff
@@ -87,15 +87,15 @@ class MacroAssemblerARM64 : public AbstractMacroAssembler<Assembler> {
-    static bool canCompact(JumpType jumpType) { return Assembler::canCompact(jumpType); }
-    static JumpLinkType computeJumpType(LinkRecord& record, const uint8_t* from, const uint8_t* to) { return Assembler::computeJumpType(record, from, to); }
-    static int jumpSizeDelta(JumpType jumpType, JumpLinkType jumpLinkType) { return Assembler::jumpSizeDelta(jumpType, jumpLinkType); }
+    ALWAYS_INLINE Vector<LinkRecord, 0, UnsafeVectorOverflow>& jumpsToLink() { return m_assembler.jumpsToLink(); }
```
Ditto for the functions these functions call - let's make them all `ALWAYS_INLINE` assuming they're all small.

ddegazio reviewed (commented):
LGTM but maybe we can expand our coverage.

MenloDorian reviewed (approved):
r=me.  Can you measure size (run `size` on the framework binary) before and after the change just so we know how much impact this has?  I don't expect a lot of impact but let's just get the data.

Commit:
Inline some functions used in copyCompactAndLinkCode
https://bugs.webkit.org/show_bug.cgi?id=266362
rdar://119627634

Reviewed by Mark Lam.

Inline some functions used in copyCompactAndLinkCode, since we want
to minimize register spilling there.

* Source/JavaScriptCore/assembler/ARM64Assembler.h:
(JSC::ARM64Assembler::buffer):
(JSC::ARM64Assembler::LinkRecord::from const):
(JSC::ARM64Assembler::LinkRecord::setFrom):
(JSC::ARM64Assembler::LinkRecord::to const):
(JSC::ARM64Assembler::LinkRecord::type const):
(JSC::ARM64Assembler::LinkRecord::linkType const):
(JSC::ARM64Assembler::LinkRecord::branchType const):
(JSC::ARM64Assembler::LinkRecord::setLinkType):
(JSC::ARM64Assembler::LinkRecord::condition const):
(JSC::ARM64Assembler::LinkRecord::is64Bit const):
(JSC::ARM64Assembler::LinkRecord::isThunk const):
(JSC::ARM64Assembler::LinkRecord::bitNumber const):
(JSC::ARM64Assembler::LinkRecord::compareRegister const):
* Source/JavaScriptCore/assembler/AssemblerBuffer.h:
(JSC::AssemblerBuffer::putInt):
(JSC::AssemblerBuffer::releaseAssemblerData):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::MacroAssemblerARM64):
(JSC::MacroAssemblerARM64::jumpsToLink):
(JSC::MacroAssemblerARM64::canCompact):
(JSC::MacroAssemblerARM64::computeJumpType):
(JSC::MacroAssemblerARM64::jumpSizeDelta):
(JSC::MacroAssemblerARM64::isCompactPtrAlignedAddressOffset):

Canonical link: https://commits.webkit.org/272034@main
